RavenDB-18144 Initial dynamic Fields support for Corax

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18144 

Started to implement this. Right now we only have simple & tuple values wired, and we can both read & write.
Need to still finish wiring everything else (Spatial, enumerable, etc).

And still need to wire the actual indexing. Adding this right now for early review.

ayende committed:
RavenDB-18144 Initial dynamic Fields support

redknightlois commented on src/Corax/IndexEntry/IndexEntryWriter.cs:
```diff
@@ -566,33 +634,27 @@ public ByteStringContext<ByteStringMemoryCache>.InternalScope Finish(out ByteStr
-
-            // Move the pointer to the end of the copied section.
-            _dataIndex += dynamicMetadataSection;
+            foreach (int fieldsLocation in _dynamicFieldsLocations!)
```
Dynamics fields pointers MUST be sorted by name.

ayende replied:
Why? 
We assume that there are going to be relatively few of them.
Likely 1 - 2 per entry.

We can just do linear scan. It is _highly_ unlikely that you'll need more.

Also remember what we'll need to use that for.
Either to list all the entries during indexing  - scan is sufficent.
Or if you sort by dynamic field (pretty rare).

redknightlois replied:
Not really, to access them we need to scan and do field name comparison. And we dont know how many dynamic fields we have on the index itself. That worst case scenario of the 100 fields in a single index is still probable. Changing that would involve very low level work to be done that is error prone, not an easy fix. I would do it correctly before we have to deal with that in production.

ayende replied:
Not really how that works. This is at the level of the _index entry_, not the global index.
And the only scenario where we actually need to do a lookup by name (vs. a scan) is for sorting by dynamic fields.

redknightlois replied:
Forgot about that. We can do the search faster with SIMD instruction if that is the case. I would still sort them though.

redknightlois commented on src/Corax/IndexEntry/IndexEntryWriter.cs:
```diff
@@ -112,6 +112,64 @@ public void Write(int field, ReadOnlySpan<byte> value)
+        WriteDynamicFieldName(fieldNameStr);
+
+        Span<byte> buffer = Buffer;
+        Unsafe.WriteUnaligned(ref buffer[_dataIndex], IndexEntryFieldType.Raw);
```
This is a `IndexEntryFieldType.Simple` not a `IndexEntryFieldType.Raw`

ayende replied:
Done

redknightlois commented on src/Corax/IndexEntry/IndexEntryWriter.cs:
```diff
@@ -566,33 +634,27 @@ public ByteStringContext<ByteStringMemoryCache>.InternalScope Finish(out ByteStr
+            foreach (int fieldsLocation in _dynamicFieldsLocations!)
+            {
+                _dataIndex += VariableSizeEncoding.Write(buffer, fieldsLocation, _dataIndex);
+            }
```
This should use the same scheme that we use for writing the known metadata table. Where depending on the size it will use fixed size values. This saves a lot of time when searching for a specific dynamic name.

ayende replied:
Why? We are not going to be doing actual searches here. And the expected scenario is to have very small number of dynamic fields.
And _searching_ for them is even more rare. Furthermore, the setup is actually more complex and space consuming than varints in most cases.

redknightlois replied:
As long as we are ok to hydrate it in a temporary, it can be done in that way.

ayende replied:
See how this is handled here, we are iterating over the values.

https://github.com/ravendb/ravendb/pull/14915/files#diff-616ac881847af034e4b7eda9043eab32ccfa0f003d7cac89a27a443fde38ddeeR483-R493

ayende committed:
RavenDB-18144 - Refactored index reading API to allow us to use the same code for both known and dynamic fields

ayende commented:
Pushed a big change now. The tests for dynamic fields aren't passing, but all the others are on my end

redknightlois commented on src/Corax/IndexEntry/IndexEntryWriter.cs:
```diff
@@ -105,13 +105,128 @@ public void Write(int field, ReadOnlySpan<byte> value)
+    {
+        using var _ = _context.From(name, ByteStringType.Immutable, out var fieldNameStr);
+        Span<byte> tmpBuffer = stackalloc byte[10];
+        int fieldLenLen = VariableSizeEncoding.Write(tmpBuffer, fieldNameStr.Length);
```
Why are we writing this into a temporary buffer?

This `stackalloc` would cause the following sequence of instructions. 
```
push 2
push 2
push 2
push 2
push 2
```
plus we are performing an encoding of the size when we are gonna discard the `tmpBuffer` later. Instead just do a guestimation of the biggest value and be done with it. 
```
var requiredSize = fieldLenLen + 2 * sizeof(long);
```

ayende replied:
I'll fix that

redknightlois commented on src/Corax/IndexEntry/IndexEntryWriter.cs:
```diff
@@ -105,13 +105,128 @@ public void Write(int field, ReadOnlySpan<byte> value)
+        
+        if (FreeSpace < requiredSize)
+            UnlikelyGrowAuxiliaryBuffer(requiredSize);
+        WriteDynamicFieldName(fieldNameStr);
```
Why are we writing the dynamic field name? Didn't you say it was not going to be stored at the entry level? I though that we would have a tree that would map those to actual `Id`s ... In fact, if we do that, then we can address directly using negative numbers instead of positive to know immediately if we have a dynamic or a static field.

redknightlois replied:
That would also save a lot of space, since we can use a `CompactTree` to store the assigned `FieldId`.

ayende replied:
That isn't a good idea, because it's *really* common to have totally unique field names.
Imagine things like: `Guid`, `DateTime`, etc as field names.

A `CompactTree` for this would require that we'll need to track usage count for dynamic fields (to allow removing it) and as well as the _inverse_ mapping, from dynamic id to the field name.

ayende committed:
RavenDB-18144 - Refactoring how we are handling indexing of terms to allow sharing code between known and dynamic indexes.

ayende committed:
RavenDB-18144
- Avoid temporary stackalloc just to compute size, use the biggest value and manage that
- Adding tests showing we can query on dynamic fields

ayende commented:
We need to think about how we can handle sorting on dynamic fields.
But first, I think we need to wire dynamic fields from RavenDB and see how that works.

redknightlois commented:
> We need to think about how we can handle sorting on dynamic fields. But first, I think we need to wire dynamic fields from RavenDB and see how that works.

Most likely what we want to do there is just sort the indexes in the metadata table. That would solve it anyways.

arekpalinski commented:
This was merged as part of https://github.com/ravendb/ravendb/pull/15000
